### ISSUE TITLE
'resolved' not test in 'npm shrinkwrap' (with replication case, node 4.4, npm 3.8.6)

### ISSUE THREAD

**Issue by sminnee** [npm/npm#12347]
I believe that this a reoccurence of #3581. I have a new case where the 'resolved' field isn't being populated in npm 3.8.6 running on node 4.4.1

Start with this package.json

    {
      "name": "tmp1",
      "version": "1.0.0",
      "description": "test",
      "main": "index.js",
      "scripts": {
        "test": "echo \"Error: no test specified\" && exit 1"
      },
      "author": "",
      "license": "ISC",
     "dependencies": {
        "blueimp-file-upload": "^6.0.3",
        "jquery-sizes": "^0.33.0"
      }
    }

Run the following commands (starting without node_modules or npm-shrinkwrap.json)

     npm install
     npm shrinkwrap

npm-shrinkwrap.json contains a sub-entry for jquery without a resolved flag

    {
      "name": "tmp1",
      "version": "1.0.0",
      "dependencies": {
        "blueimp-canvas-to-blob": {
          "version": "3.3.0",
          "from": "blueimp-canvas-to-blob@>=2.0.3",
          "resolved": "https://registry.npmjs.org/blueimp-canvas-to-blob/-/blueimp-canvas-to-blob-3.3.0.tgz"
        },
        "blueimp-file-upload": {
          "version": "6.9.7",
          "from": "blueimp-file-upload@>=6.0.3 <7.0.0",
          "resolved": "https://registry.npmjs.org/blueimp-file-upload/-/blueimp-file-upload-6.9.7.tgz"
        },
        "blueimp-load-image": {
          "version": "2.6.1",
          "from": "blueimp-load-image@>=1.2.1",
          "resolved": "https://registry.npmjs.org/blueimp-load-image/-/blueimp-load-image-2.6.1.tgz"
        },
        "blueimp-tmpl": {
          "version": "3.4.0",
          "from": "blueimp-tmpl@>=2.1.0",
          "resolved": "https://registry.npmjs.org/blueimp-tmpl/-/blueimp-tmpl-3.4.0.tgz"
        },
        "jquery": {
          "version": "2.2.3",
          "from": "jquery@>=1.6.0",
          "resolved": "https://registry.npmjs.org/jquery/-/jquery-2.2.3.tgz"
        },
        "jquery-sizes": {
          "version": "0.33.0",
          "from": "jquery-sizes@>=0.33.0 <0.34.0",
          "resolved": "https://registry.npmjs.org/jquery-sizes/-/jquery-sizes-0.33.0.tgz",
          "dependencies": {
            "jquery": {
              "version": "2.2.3",
              "from": "jquery@latest"
            }
          }
        },
        "jquery.ui.widget": {
          "version": "1.10.3",
          "from": "jquery.ui.widget@>=1.8.0",
          "resolved": "https://registry.npmjs.org/jquery.ui.widget/-/jquery.ui.widget-1.10.3.tgz"
        }
      }
    }

This causes issues if you attempt to install from shrinkwrap and re-wrap

    rm -rf node_modules
    npm install 
    npm shrinkwrap

The error is:

	npm ERR! Darwin 15.4.0
	npm ERR! argv "/usr/local/Cellar/node4-lts/4.4.1/bin/node" "/usr/local/bin/npm" "shrinkwrap"
	npm ERR! node v4.4.1
	npm ERR! npm  v3.8.6

	npm ERR! Problems were encountered
	npm ERR! Please correct and try again.
	npm ERR! invalid: have jquery@2.2.3 (expected: latest) /Users/sminnee/Projects/tmp/tmp1/node_modules/jquery-sizes/node_modules/jquery
	npm ERR! extraneous: jquery@2.2.3 /Users/sminnee/Projects/tmp/tmp1/node_modules/jquery-sizes/node_modules/jquery
	npm ERR! 
	npm ERR! If you need help, you may report this error at:
	npm ERR!     <https://github.com/npm/npm/issues>

	npm ERR! Please include the following file with any support request:
	npm ERR!     /Users/sminnee/Projects/tmp/tmp1/npm-debug.log

**Comment by nevir** [comment]
Experiencing this as well, via npm `3.8.3` on node `5.9.1`

**Comment by sminnee** [comment]
So, related to this issue is that `node_modules\jquery-sizes\node_modules\jquery\package.json` doesn't have a `_resolved` entry. My assumption is that the best way to fix this bug is to ensure that `_resolved` is populated in this file.

This fix corrects that issue, however, it will impact a lot of use-cases as it's changing the output of a core support library.

	diff --git a/lib/fetch-package-metadata.js b/lib/fetch-package-metadata.js
	index 9e62cac..7177b8e 100644
	--- a/lib/fetch-package-metadata.js
	+++ b/lib/fetch-package-metadata.js
	@@ -111,9 +111,6 @@ function fetchNamedPackageData (dep, next) {
	       npm.registry.get(url, {auth: auth}, pulseTillDone('fetchMetadata', iferr(next, pickVersionFromRegistryDocument)))
	     }
	     function returnAndAddMetadata (pkg) {
	-      delete pkg._from
	-      delete pkg._resolved
	-      delete pkg._shasum
	       next(null, pkg)
	     }
	     function pickVersionFromRegistryDocument (pkg) {

It begs the question, what is the reason for excluding this meta-data from the results of fetch-package-metadata? Does it cause problems down the line?

**Comment by sminnee** [comment]
OK I found a better fix and raised a PR at #12426.

**Comment by sminnee** [comment]
OK @iarna I thought that PR fixed this but I still get another issue.

The bug occurs when you re-shrinkwrap modules installed from a shrinkwrap-provided installation. In other words, this sequence of commands:

 * npm install
 * npm shrinkwrap
 * rm -rf node_modules
 * npm install
 * npm shrinkwrap

Gives this error:

	npm WARN tmp1@1.0.0 No repository field.
	npm ERR! Darwin 15.4.0
	npm ERR! argv "/usr/local/Cellar/node4-lts/4.4.1/bin/node" "/<path>/npm/bin/npm-cli.js" "shrinkwrap"
	npm ERR! node v4.4.1
	npm ERR! npm  v3.8.3

	npm ERR! Problems were encountered
	npm ERR! Please correct and try again.
	npm ERR! invalid: have jquery@2.2.3 (expected: latest) /<path>/node_modules/jquery-sizes/node_modules/jquery
	npm ERR! extraneous: jquery@2.2.3 /<path>/node_modules/jquery-sizes/node_modules/jquery
	npm ERR! 
	npm ERR! If you need help, you may report this error at:
	npm ERR!     <https://github.com/npm/npm/issues>

	npm ERR! Please include the following file with any support request:
	npm ERR!     /<path>/npm-debug.log

I think that the cause is that `_from` doesn't match the spec of the original `jquery-sizes` dependency.

**Comment by sminnee** [comment]
Scratch that—this bug is why the removal of the conditional was important! ;-)

**Comment by sminnee** [comment]
(edited the previous comment after getting some more data)

**Comment by iarna** [comment]
I think it's actually that it's looking at `_requested`, when for shrinkwraps at least, `_from` would be better.

Specifically this is happening when you shrinkwrap packages who depend on tags.

**Comment by iarna** [comment]
I should have a PR to fix this shortly.

**Comment by iarna** [comment]
I think this will fix it:
```
diff --git a/lib/install/deps.js b/lib/install/deps.js
index e154035..0cdd9e6 100644
--- a/lib/install/deps.js
+++ b/lib/install/deps.js
@@ -72,7 +72,7 @@ function doesChildVersionMatch (child, requested, requestor) {
     if (childReq.rawSpec === requested.rawSpec) return true
     if (childReq.type === requested.type && childReq.spec === requested.spec) return true
   }
-  if (!registryTypes[requested.type]) return requested.rawSpec === child.package._from
+  if (!registryTypes[requested.type]) return requested.raw === child.package._from
   return semver.satisfies(child.package.version, requested.spec)
 }
```

`rawSpec` is the version part of `package@version` before it undergoes normalization. By contrast, `raw` is the entire specifier. The existing check worked when the entire specifier was just a URL to a tarball, but it didn't work for tags like `latest`.

**Comment by sminnee** [comment]
Yeah I had a look yesterday and that was my conclusion too.

Related, I also found issues with `add-remote-git`, see https://gist.github.com/fc99b537e6ad18bfd6032cc3954172f5 which helps to fix #11736

The common pattern is that `_from` needs to be consistently set to the package dependency in the manner specified by the requester, rather than any kind of resolution of what that maps to.
